### PR TITLE
Streaming Timeseries example

### PR DESCRIPTION
**Pull request by droumis**
Example to demonstrate streaming multichannel timeseries data. 

Can stream live from CPUs as a way to confirm that the streaming machinery is working.

Can also stream from an LSL stream (mock from EEG file) which is used throughout research for signal acquisition.

streaming_header

image

### CODE REVIEW

**Comment by amaloney** [comment]
@droumis would you mind describing in more detail what "smoother streaming updates" means to you?

**Comment by droumis** [comment]
@amaloney. Minimize latency between patch updates. Of course, we can't go below the data sampling interval, but I think HoloViews will probably be more of the limiting factor that might cap us at updating the viewport once every 10-20 ms or so at max (just a guess). This might involve buffering a smaller amount of data for smoother perceived interpolation without noticeable lag and/or ensuring the buffered dataframe range is consistent across the channels/curves so as to maybe minimize delays between data arrival and plotting.

**Comment by maximlt** [comment]
@amaloney the test pipeline is broken due to some unrelated issue for which a fix has already been merged on `main`, so it should go green whenever `main` is merged onto this branch.

I also saw you applied black/isort to the code in this PR. You may be wondering why we don't do that automatically. If not, ignore me :D But if so, the main reason is that we want to let the example author format the code the way they prefer, which can sometimes be more readable than when done automatically by a tool. In the end, when I review an example the only things I care about about code formatting are that it is not too weird and that the lines aren't too wide (painful to read on mobile).

**Comment by amaloney** [comment]
Thanks for the clarification @droumis I'll work on that today.

Also, thanks for the heads-up @maximlt. Your points about `black` and `isort` make perfect sense. The point of examples are about the contribution.

**Comment by jbednar** [comment]
Really cool example!

File: streaming_timeseries/streaming_timeseries.ipynb [hunk 1]
```diff
@@ -0,0 +1,948 @@
...
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "class StreamingApp(pn.viewable.Viewer):\n",
+    "    def __init__(\n",
+    "        self,\n",
+    "        data_sources: list,\n",
+    "        notebook: bool = True,\n",
+    "        buffer_length_samples: int = 5000,\n",
```
Comment (maximlt) [inline streaming_timeseries/streaming_timeseries.ipynb:436]: Since this is a Parameterized class these could be declared as Parameters.

**Review by maximlt — COMMENTED** [review]
Nice exampled indeed :)

I have to main pieces of feedback:
- It'd be nice to have a bit more of neuroscience-y context and narrative. What are these EEG traces I'm looking at? What does the channel positions plot represent?
- The `StreamingApp` class is quite long and a bit hard to follow. It'd be nice if:
  - A few of the main concepts could be explained beforehand (`hv.DynamicMap`, `hv.streams.Buffer`,  `pn.state.add_periodic_callback`, ` pn.io.unlocked`).
  - The class was refactored a bit. For example, the plot methods (`create_{}`, `bare_{}`) could be defined as functions before the class is defined, and could be called once separately to show what output they generate. The methods in the class would call these functions. Of course, only if this refactor isn't too much work!

**Comment by maximlt** [comment]
@amaloney FYI in https://github.com/holoviz-topics/examples/pull/469/commits/3992c20cb0bb8b3020c35b94fcdbdd3e44181d7b I cleared the notebook cell outputs. On the other HoloViz repos this is usually done automatically via a pre-commit hook Simon maintains (https://github.com/hoxbro/clean_notebook). This repo is different as it needs to allow some example notebooks to be committed with their full output, to deal with examples that can't run on the CI (need GPU, ultra-large datasets, too long computational time, etc.). So, we just clear the standard notebooks manually before committing them. If not, the test suite on the CI will detect they contain outputs and fail.

File: streaming_timeseries/streaming_timeseries.ipynb [hunk 2]
```diff
@@ -0,0 +1,948 @@
...
+    "            min_height=min_height,\n",
+    "            responsive=True,\n",
+    "        )\n",
+    "        return curve\n",
+    "\n",
+    "    def empty_stream_plot(self) -> hv.NdOverlay:\n",
+    "        empty_curves = {\n",
+    "            ch: hv.Curve([]).relabel(ch).opts(\n",
+    "                subcoordinate_y=True,\n",
+    "                color=self.color_mapping.get(ch, None),\n",
```
[the remark this answers is not in the document]
Comment (amaloney) [reply streaming_timeseries/streaming_timeseries.ipynb:738]: To me, a hover allows a user to acquire more information about a point in a plot. The only time the plot is static enough for hovers to show is when the data is paused, but pausing and inspecting data in this example is not really the purpose of the example.

I could be wrong in this thought process though.

File: streaming_timeseries/streaming_timeseries.ipynb [hunk 3]
```diff
@@ -0,0 +1,948 @@
...
+    "            axiswise=True,\n",
+    "            height=300,\n",
+    "            responsive=True,\n",
+    "            shared_axes=False,\n",
+    "            title=\"Channel Position\",\n",
+    "        )\n",
+    "        return points\n",
+    "\n",
+    "    def __panel__(self) -> pn.Row | pn.Column:\n",
+    "        if self.notebook:\n",
```
[the remark this answers is not in the document]
Comment (amaloney) [reply streaming_timeseries/streaming_timeseries.ipynb:764]: Is your suggestion considered "best practice"? I ask because I don't know, but what you suggest makes sense to me. If you have examples I could reference, I'd greatly appreciate them.

File: streaming_timeseries/streaming_timeseries.ipynb [hunk 4]
```diff
@@ -0,0 +1,948 @@
...
+    "            self.pause_stream()\n",
+    "        elif event.new == \"Stop\":\n",
+    "            self.stop_stream()\n",
+    "\n",
+    "    def create_layout(self) -> None:\n",
+    "        # Initialize main visualization panes with empty plots\n",
+    "        self.position_pane = pn.pane.HoloViews(self.bare_pos_plot())\n",
+    "        self.main_streaming_pane = pn.pane.HoloViews(self.bare_stream_plot())\n",
+    "\n",
+    "        if self.notebook:\n",
```
[the remark this answers is not in the document]
Comment (amaloney) [reply streaming_timeseries/streaming_timeseries.ipynb:537]: I think this is a great suggestion. I didn't even know it existed, but it's definitely in the documentation https://panel.holoviz.org/api/state.html. Do you have examples that use this feature? If so, I think it would be nice to show some links to it in this notebook.

**Comment by amaloney** [comment]
@maximlt add other items as you see fit. I don't think the following are too time consuming to implement.

**`TODO`**

- [ ] add neuroscience-y context and narrative
  - [ ] What are these EEG traces I'm looking at?
  - [ ] What does the channel positions plot represent?
- [ ] explain the following components of the streaming app
  - [ ] `hv.DynamicMap`
  - [ ] `hv.streams.Buffer`
  - [ ] `pn.state.add_periodic_callback`
  - [ ] `pn.io.unlocked`
- [ ] refactor streaming app
  - [ ] define `create_{}` and `bare_{}` before the class, and call separately to show what output

File: streaming_timeseries/anaconda-project.yml [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+description: Streaming display of multichannel timeseries data.
+
+examples_config:
+  created: 2024-12-06
+  maintainers:
+  - "droumis"
+  labels:
+  - "panel"
+  - "holoviews"
+  categories:
```
Comment (maximlt) [inline streaming_timeseries/anaconda-project.yml:11]: I've recently updated all the labels and added `bokeh` when hvplot/holoviews generate bokeh plots.

```suggestion
  - "holoviews"
  - "bokeh"
  categories:
```

File: streaming_timeseries/streaming_timeseries.ipynb (hunk 2 above)
[the remark this answers is not in the document]
Comment (maximlt) [reply streaming_timeseries/streaming_timeseries.ipynb:738]: I sort of naturally wanted to see the values, at least for the CPU example, as without it it's hard to see if you're close to 100%.

image

HoloViews [docs](https://holoviews.org/gallery/demos/bokeh/multichannel_timeseries_viewer.html#visualizing-eeg-data) say (worth checking :) ):

> By default, all the curves, including across data groups, have the same y-axis range that is computed from the min and max across all channels.

Setting `.opts(ylim=(0, 100))` would be more appropriate for the CPU example? I don't know setting `ylim` for the EEG traces would make sense or not. Maybe a data source could have a `ylim` property whose value would be used to set `opts(ylim=...)`, if the property doesn't return `None`.

File: streaming_timeseries/streaming_timeseries.ipynb (hunk 4 above)
[the remark this answers is not in the document]
Comment (maximlt) [reply streaming_timeseries/streaming_timeseries.ipynb:537]: I use `pn.state.served` quite a lot in some code bases for the code to follow a different branch, usually to distinguish between the app being served and the test suite running.

Though I just remembered there were some subtleties with it (see more in https://github.com/holoviz/panel/issues/5623). Since no other example use it and it's yet another thing to explain in a rather complex example, I'd say forget about my suggestion and let's stick with the more explicit `notebook` attribute?

File: streaming_timeseries/streaming_timeseries.ipynb (hunk 3 above)
[the remark this answers is not in the document]
Comment (maximlt) [reply streaming_timeseries/streaming_timeseries.ipynb:764]: I don't think we have an example with a `__panel__` view that returns either a layout or a template based on the context. Usually, they build a bunch of separate components, put them together in a layout for the purpose of the notebook, and then arrange them in a template (not displayed in the notebook) to make the notebook panel-servable.
In this case, I don't see the need for two attributes.

**Comment by maximlt** [comment]
> @maximlt so don't clear the outputs after addressing the PR comments?

The other way around, please always clear the outputs unless it's a very special example :)

> explain the following components of the streaming app

Perhaps a good approach would be to come up with a minimalist example (e.g. 1 curve from random data) using these 4 building blocks (Buffer, DynamicMap, etc.). I'm personally curious about `pn.io.unlocked` which I've never used. Oh and I know I'm asking for quite some work so feel free to push against it if you think it's not worth it.

File: streaming_timeseries/streaming_timeseries.ipynb (hunk 4 above)
[the remark this answers is not in the document]
Comment (amaloney) [reply streaming_timeseries/streaming_timeseries.ipynb:537]: Sure, although I would personally like more info into its use as it looks like it could be a great flag to use. I'll see if I can put something on our calendar.

File: streaming_timeseries/streaming_timeseries.ipynb (hunk 3 above)
[the remark this answers is not in the document]
Comment (amaloney) [reply streaming_timeseries/streaming_timeseries.ipynb:764]: Oddly, if you change the app to your suggested method, then the stand-alone version breaks. I'm leaving it as is for now.

File: streaming_timeseries/streaming_timeseries.ipynb [hunk 6]
```diff
@@ -0,0 +1,948 @@
...
+    "\n",
+    "In the class above, we created a condition when `notebook=False` that adds our plot to the `main`\n",
+    "area of a Panel Template component and puts the widgets in the template's `sidebar`. All we need to\n",
+    "do now is create an instance and mark it as `servable`.\n",
+    "\n",
+    "To launch the standalone app, activate the same conda environment and run\n",
+    "`panel serve <path-to-this-file> --show` in the command line. This will open the application in a\n",
+    "browser window.\n",
+    "\n",
+    "<div class=\"admonition alert alert-warning\">\n",
```
[the remark this answers is not in the document]
Comment (amaloney) [reply streaming_timeseries/streaming_timeseries.ipynb:867]: I did not see an issue running both simultaneously when the data sources were from CPU percentages. The app in the notebook broke when switching to the EEG data, when the EEG data was being displayed in the stand-alone. I'll make a note of this in the nb.

File: streaming_timeseries/streaming_timeseries.ipynb (hunk 1 above)
Comment (amaloney) [reply streaming_timeseries/streaming_timeseries.ipynb:436]: correct, but I'm going to leave as is, unless there is a reason I cannot think of as to why they should be Parameters, and you can help me understand why they should be.

Comment (maximlt) [reply streaming_timeseries/streaming_timeseries.ipynb:436]: I wouldn't have asked that if the class wasn't inheriting from `param.Paramaterized` but it does (through `Viewer`). It seems to me that in HoloViz land it is standard practice in this case to have the `__init__` keywords defined as `Parameter`s. Happy to hear from @droumis too!

**Comment by droumis** [comment]
> Perhaps a good approach would be to come up with a minimalist example (e.g. 1 curve from random data) using these 4 building blocks (Buffer, DynamicMap, etc.). 

I opted to add a minimap streaming app towards the beginning with some random data. I also included extensive explanation later on about each of the implementation steps in question. 

I am choosing **not** to refactor the app to strip out the `create_` and `bare_` methods. Not only is it too much work right now, but I think the primary use case of this will be as a standalone app and in such a case, I want the class to be more like how it would coded as a standalone python script. I think the addition of the minimal app fills the need for adding some incremental steps to the workflow tutorial.

File: streaming_timeseries/streaming_timeseries.ipynb (hunk 1 above)
Comment (droumis) [reply streaming_timeseries/streaming_timeseries.ipynb]: ok, I refactored

File: streaming_timeseries/streaming_timeseries.ipynb [hunk 7]
```diff
@@ -0,0 +1,948 @@
...
+   "metadata": {
+    "editable": true,
+    "slideshow": {
+     "slide_type": ""
+    },
+    "tags": []
+   },
+   "outputs": [],
+   "source": [
+    "nb_app"
```
[the remark this answers is not in the document]
Comment (droumis) [reply streaming_timeseries/streaming_timeseries.ipynb:937]: it might be cool if we could have it trigger the cpu stream for a bit but only in the website build and otherwise hide the code from the tutorial notebook. But I don't have time to figure that out and also idk what it might even show for CPU usage on a CI system. So, dropping given urgency of initial release.. maybe we can come back to this later

File: streaming_timeseries/streaming_timeseries.ipynb (hunk 4 above)
[the remark this answers is not in the document]
Comment (droumis) [reply streaming_timeseries/streaming_timeseries.ipynb:653]: yea, using a notebook flag is super hacky and I'd much prefer if there was a way to detect with pn.state.served, however, given the timeline, let's just come back to this in a later PR

File: streaming_timeseries/streaming_timeseries.ipynb (hunk 3 above)
[the remark this answers is not in the document]
Comment (droumis) [reply streaming_timeseries/streaming_timeseries.ipynb:889]: oof yea idk what's going on here. We should come back to this after initial release and standardize on a best practice and figure out why it's not working here in standalone

**Comment by droumis** [comment]
- [ ] add event detection
- [ ] add memory load in side plot for cpu stream
- [ ] add image stream with pos estimation in additional side plot

**Review by maximlt — APPROVED** [review]
LGTM!

File: streaming_timeseries/streaming_timeseries.ipynb (hunk 7 above)
[the remark this answers is not in the document]
Comment (maximlt) [reply streaming_timeseries/streaming_timeseries.ipynb:935]: Just realized that the `await` here works fine in the context of a notebook runtime (where there's already an event loop), but I'm not sure Python would even let the app being served (SyntaxError?).

An other approach inspired by the simpler example Demetris added would be to instantiate the class with a `timeout` and pass it down to the callback.
